File: GuildTools/ClientApp/src/app/permissions/permissions-order.ts
import { GuildProfilePermissionLevel } from 'app/auth/auth.service';

interface SortablePermissionLevel {
  level: GuildProfilePermissionLevel;
  order: number;
}

export class PermissionsOrder {

  private static _instance: PermissionsOrder = null;

  private static get staticInstance() {
    if (PermissionsOrder._instance === null){
      PermissionsOrder._instance = new PermissionsOrder();
    }

    return PermissionsOrder._instance;
  }

  private orderMap = new Map<GuildProfilePermissionLevel, number>();
  private labelMap = new Map<GuildProfilePermissionLevel, string>();

  constructor() {
    this.orderMap.set(GuildProfilePermissionLevel.Visitor, 1);
    this.orderMap.set(GuildProfilePermissionLevel.Member, 2);
    this.orderMap.set(GuildProfilePermissionLevel.Officer, 3);
    this.orderMap.set(GuildProfilePermissionLevel.Admin, 4);

    this.labelMap.set(GuildProfilePermissionLevel.Visitor, 'Visitor');
    this.labelMap.set(GuildProfilePermissionLevel.Member, 'Member');
    this.labelMap.set(GuildProfilePermissionLevel.Officer, 'Officer');
    this.labelMap.set(GuildProfilePermissionLevel.Admin, 'Admin');
  }

  public static GreaterThanOrEqual(target: GuildProfilePermissionLevel, compareAgainst: GuildProfilePermissionLevel): boolean {
    return PermissionsOrder.staticInstance.orderMap.get(target) >= PermissionsOrder.staticInstance.orderMap.get(compareAgainst);
  }

  public static LessThanOrEqual(target: GuildProfilePermissionLevel, compareAgainst: GuildProfilePermissionLevel): boolean {
    return PermissionsOrder.staticInstance.orderMap.get(target) <= PermissionsOrder.staticInstance.orderMap.get(compareAgainst);
  }

  public static GetOrderedPermissions(): Array<GuildProfilePermissionLevel> {
    const sortableArray = new Array<SortablePermissionLevel>();

    for (const [permission, order] of PermissionsOrder.staticInstance.orderMap.entries()) {
      sortableArray.push({
        level: permission,
        order: order
      });
    }

    sortableArray.sort((a, b) => {
      return a.order - b.order;
    });

    return sortableArray.map(x => x.level);
  }

  public static GetPermissionLabel(level: GuildProfilePermissionLevel): string {
    return PermissionsOrder.staticInstance.labelMap.get(level);
  }
}

File: GuildTools/Permissions/PermissionsOrder.cs
using GuildTools.EF.Models.Enums;
using System;
using System.Collections.Generic;
using System.Linq;
using System.Threading.Tasks;

namespace GuildTools.Permissions
{
    public class PermissionsOrder
    {
        private static Dictionary<GuildProfilePermissionLevel, int> PermissionOrderMap =
            new Dictionary<GuildProfilePermissionLevel, int>()
                {
                    { GuildProfilePermissionLevel.Visitor, 1 },
                    { GuildProfilePermissionLevel.Member, 2 },
                    { GuildProfilePermissionLevel.Officer, 3 },
                    { GuildProfilePermissionLevel.Admin, 4 }
                };

        public static int GetPermissionOrder(GuildProfilePermissionLevel level)
        {
            return PermissionOrderMap[level];
        }

        public static bool GreaterThanOrEqual(GuildProfilePermissionLevel target, GuildProfilePermissionLevel compareAgainst)
        {
            return PermissionsOrder.PermissionOrderMap[target] >= PermissionsOrder.PermissionOrderMap[compareAgainst];
        }

        public static int Order(GuildProfilePermissionLevel level)
        {
            return PermissionsOrder.PermissionOrderMap[level];
        }
    }
}
